### PR TITLE
Machine module

### LINKED ISSUE #3: Implement machines module
Module for `machines` should contain such components:
- `machines.controller.go`
- `machines.service.go`
- `machines.repository.go`
- `machines.entity.go`

Machines entity:
- `id`: `uint32`
- `name`: `string`
- `cpuCount`: `uint8`
- `totalDiskSpace`: `uint64` (in bytes)

Machines service:
- `readAll`
- `findById`
- `update`

Machines controller:
- GET: `/machines`

### PR DESCRIPTION
**Pull request by lightballer**
Closes #3

### CODE REVIEW

File: server/machines/machines.controller.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package machines
```
Comment (yehorbk) [inline server/machines/machines.controller.go:1]: Refactor file using `mux`. Take a look on [DisksController](https://github.com/multipleton/sa-3/pull/11/files#diff-00562082e2c7f0e951a98ef7aae5d04b5ec2848e38270f80f388345e861554f6)

File: server/machines/machines.controller.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+import (
+	"encoding/json"
+	"net/http"
+)
+
+type HttpHandlerFunc http.HandlerFunc
```
Comment (yehorbk) [inline server/machines/machines.controller.go]: Does not match specification #3. It should be `MachinesController`

File: server/machines/machines.repository.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	"database/sql"
+	"fmt"
+)
+
+type MachinesRepository struct {
+	Db *sql.DB
+}
+
+func (mr *MachinesRepository) GetAll() ([]Machine, error) {
+	rows, err = r.DB.Query("SELECT id, name, cpu_count, total_disk_space FROM machines")
```
Comment (olegmak02) [inline server/machines/machines.repository.go]: you wrote 'r.DB.Query', but here should be 'mr.DB.Query'

File: server/machines/machines.controller.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+		if r.Method == "GET" {
+			handleListMachines(rw, ms)
+		} else {
+			rw.WriteHeader(http.StatusMethodNotAllowed)
+		}
+	}
+}
+
+func handleListMachines(rw http.ResponseWriter, ms MachinesService) {
+	rw.Header().Set("content-type", "application/json")
```
Comment (yehorbk) [inline server/machines/machines.controller.go]: `git cherry-pick e0ef103bc97fdbceb6affc6e1b3fda94114bccc0` and use `Respond` function

File: server/machines/machines.controller.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+	res, err := ms.ReadAll()
+	if err != nil {
+		rw.WriteHeader(404)
+		_ := json.NewEncoder(rw).Encode("cannot read all machines")
+		return
```
Comment (yehorbk) [inline server/machines/machines.controller.go]: If no entries - return `[]`, not `error`

File: server/machines/machines.entity.go [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package machines
+
+type Machine struct {
```
Comment (yehorbk) [inline server/machines/machines.entity.go]:
```suggestion
type MachinesEntity struct {
```

File: server/machines/machines.repository.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
+package machines
+
+import (
+	"database/sql"
+	"fmt"
+)
+
+type MachinesRepository struct {
+	Db *sql.DB
```
Comment (yehorbk) [inline server/machines/machines.repository.go]: Should be private
```suggestion
	db *sql.DB
```

File: server/machines/machines.repository.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+func (mr *MachinesRepository) GetAll() ([]Machine, error) {
```
Comment (yehorbk) [inline server/machines/machines.repository.go]:
```suggestion
func (mr *MachinesRepository) ReadAll() ([]MachinesEntity, error) {
```

File: server/machines/machines.repository.go (hunk 3 above)
Comment (lightballer) [reply server/machines/machines.repository.go]: yep, thanks

File: server/machines/machines.repository.go [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+		}
+		res = append(res, &m)
+	}
+	if res == nil {
+		res = make([]*Machine, 0)
+	}
+	return res, nil
+}
+
+func (mr *MachinesRepository) GetOne(id string) (Machine, error) {
```
Comment (yehorbk) [inline server/machines/machines.repository.go]: `id` should be `uint32`
```suggestion
func (mr *MachinesRepository) FindById(id uint32) (MachinesEntity, error) {
```

File: server/machines/machines.repository.go [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+	query := fmt.Sprintf("SELECT id, name, cpu_count, total_disk_space FROM machines WHERE id=%s", id)
+	row, err = r.DB.Query(query)
```
Comment (yehorbk) [inline server/machines/machines.repository.go]: Use `QueryRow` function

File: server/machines/machines.repository.go [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+	defer row.Close()
+	var m Machine
+	if err := row.Scan(&m.Id, &m.Name, &m.CpuCount, &m.TotalDiskSpace); err != nil {
+		return nil, err
+	}
+	res = append(res, &m)
+	return res, nil
+}
+
+func (mr *MachinesRepository) UpdateOne(id string, machine Machine) (Machine, error) {
```
Comment (yehorbk) [inline server/machines/machines.repository.go]:
```suggestion
func (mr *MachinesRepository) Update (entity MachinesEntity) (MachinesEntity, error) {
```

File: server/machines/machines.repository.go [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+	query := fmt.Sprintf("UPDATE machines SET total_disk_space=%s WHERE id=%s RETURNING id, name, cpu_count, total_disk_space", machine.TotalDiskSpace, id)
+	row, err = r.DB.Query(query)
```
Comment (yehorbk) [inline server/machines/machines.repository.go]: Use `Exec` function

File: server/machines/machines.service.go [hunk 13]
```diff
@@ -0,0 +1,33 @@
+package machines
+
+type MachinesService struct {
+	Mr MachinesRepository
```
Comment (yehorbk) [inline server/machines/machines.service.go]: Should be private
```suggestion
	machinesRepository MachinesRepository
```

File: server/machines/machines.service.go [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+
+func (ms *MachinesService) FindById(id string) (Machine, error) {
+	machine, err := ms.Mr.GetOne(id)
+	if err != nil {
+		return nil, err
+	}
+	return machine, nil
+}
+
+func (ms *MachinesService) Update(id string, machine Machine) (Machine, error) {
```
Comment (yehorbk) [inline server/machines/machines.service.go]:
```suggestion
func (ms *MachinesService) Update(id uint32, machine MachinesEntity) (MachinesEntity, error) {
```

File: server/machines/machines.service.go [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+	machine, err := ms.Mr.GetOne(id)
+	if err != nil {
+		return nil, err
+	}
+	updatedMachine, err := ms.Mr.UpdateOne(id, machine)
+	if err != nil {
+		return nil, err
+	}
+	return updatedMachine, nil
```
Comment (yehorbk) [inline server/machines/machines.service.go]: 👍

**Review by yehorbk — CHANGES_REQUESTED** [review]
- Make refactoring following suggested changes/comments
- MachinesController should be refactored with `mux` to suit #13 
- It is not really necessary to fix names in `MachinesRepository`, but to make code consistent you can do that

File: server/machines/machines.controller.go (hunk 2 above)
Comment (lightballer) [reply server/machines/machines.controller.go]: why?

File: server/machines/machines.controller.go (hunk 5 above)
Comment (lightballer) [reply server/machines/machines.controller.go]: if no entries returns `[]`

File: server/machines/machines.repository.go (hunk 12 above)
Comment (lightballer) [reply server/machines/machines.repository.go]: why

File: server/machines/machines.controller.go (hunk 5 above)
Comment (yehorbk) [reply server/machines/machines.controller.go]: If you have error then return 500, not 404

File: server/machines/machines.repository.go (hunk 12 above)
Comment (yehorbk) [reply server/machines/machines.repository.go]: It's better to return value with findById

Comment (lightballer) [reply server/machines/machines.repository.go]: I return value anyway, what's wrong?

File: server/machines/machines.controller.go [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+	"github.com/multipleton/sa-3/utils"
+	"net/http"
+)
+
+type MachinesController struct {
+	ms *MachinesService
+}
+
+func (mc *MachinesController) HandleRoutes(router *mux.Router) {
+	router.HandleFunc("/machines/", mc.ServeAllMachines).Methods("GET")
```
Comment (yehorbk) [inline server/machines/machines.controller.go]:
```suggestion
	router.HandleFunc("/machines", mc.ServeAllMachines).Methods("GET")
```

File: server/machines/machines.controller.go [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+func (mc *MachinesController) ServeAllMachines(rw http.ResponseWriter, r *http.Request) {
+	res, err := mc.ms.ReadAll()
+	if err != nil {
+		utils.Respond(rw, 500, errors.New("internal server error"))
```
Comment (yehorbk) [inline server/machines/machines.controller.go]:
```suggestion
		utils.Respond(rw, 500, err)
```

File: server/machines/machines.controller.go [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+		return
+	}
+	if err := json.NewEncoder(rw).Encode(res); err != nil {
+		utils.Respond(rw, 500, errors.New("internal server error"))
```
Comment (yehorbk) [inline server/machines/machines.controller.go]:
```suggestion
		utils.Respond(rw, 500, err)
```

**Review by yehorbk — CHANGES_REQUESTED** [review]

**Review by yehorbk — APPROVED** [review]

**Pull request merged**